[docs] Add transitions customization page

mnajdova opened the pull request:
Docs page: https://deploy-preview-21456--material-ui.netlify.app/customization/transitions/

Closes https://github.com/mui-org/material-ui/issues/16928

oliviertassinari commented on docs/src/pages/customization/transitions/TransitionHover.tsx:
```diff
@@ -0,0 +1,24 @@
+      duration: theme.transitions.duration.standard,
+    }),
+    '&:hover': {
+      backgroundColor: deepPurple[900],
```
I think hat we could make it more visible, something like this maybe https://tailwindcss.com/docs/transition-property?

mnajdova replied:
Improved the example 👍

oliviertassinari commented on docs/src/pages/customization/transitions/transitions.md:
```diff
@@ -0,0 +1,33 @@
+
+<p class="description">Use the <code>theme.transitions.create()</code> helper to create consistent transitions on the elements of your UI.</p>
+
+{{"demo": "pages/customization/transitions/TransitionHover.js"}}
```
I don't think that the inline preview is of any value. We can disable it to reduce noise.

```suggestion
{{"demo": "pages/customization/transitions/TransitionHover.js", "defaultCodeOpen": false}}
```

oliviertassinari replied:
Or we could do the opposite, open it completely
```suggestion
{{"demo": "pages/customization/transitions/TransitionHover.js", "defaultCodeOpen": true}}
```

mnajdova replied:
Let me change the example to see how long it would be and I will check what will look better

oliviertassinari commented on docs/src/pages/customization/transitions/transitions.md:
```diff
@@ -0,0 +1,33 @@
+# Transitions
+
+<p class="description">Use the <code>theme.transitions.create()</code> helper to create consistent transitions on the elements of your UI.</p>
```
the generated HTML is:

```html
<meta name="description" content="Use the &lt;code&gt;theme.transitions.create()&lt;/code&gt; helper to create consistent transitions on the elements of your UI."/>
```

I'm curious to see what it looks like once indexed. We will see :)

mnajdova commented:
> I think that it would be great to add an ## API section to document the different options.

@oliviertassinari is there some automatic ways that we can use here for the API, or should I create it manually? 

>  It might also be great to have a demo with styled-components, similar to https://material-ui.com/guides/interoperability/#theme.

I converted the example to use `styled-components`. Do we want to have one using `makeStyles`, or is this one enough?

oliviertassinari commented:
> is there some automatic ways that we can use here for the API, or should I create it manually?

I fear we don't have any automated scripts for it :/. The pages in https://material-ui.com/customization/breakpoints/#api were created by hand. 

> I converted the example to use styled-components. Do we want to have one using makeStyles, or is this one enough?

I guess styled-components is enough? No strong preference.

mnajdova commented:
> I fear we don't have any automated scripts for it :/. The pages in https://material-ui.com/customization/breakpoints/#api were created by hand.

This is perfect as an exmaple, didn't want to create something completely different :) 

> I guess styled-components is enough? No strong preference.

I think so, let's keep just that one 👍

oliviertassinari commented on docs/src/pages/customization/transitions/transitions.md:
```diff
@@ -0,0 +1,50 @@
+# Transitions
+
+<p class="description">Use the <code>theme.transitions.create()</code> helper to create consistent transitions on the elements of your UI.</p>
+
+{{"demo": "pages/customization/transitions/TransitionHover.js", "defaultCodeOpen": false}}
```
I think that it could be interesting to have a visual example people can directly copy & paste.

```suggestion
```jsx
theme.transitions.create(['background-color', 'transform']);
```

{{"demo": "pages/customization/transitions/TransitionHover.js", "defaultCodeOpen": false}}
```

mnajdova replied:
Sry I don’t understand what is the suggested change here. Should I add code snippet after the example?

mnajdova replied:
Ah I see the code snippet is added before the example :)

oliviertassinari commented on docs/src/pages/customization/transitions/transitions.md:
```diff
@@ -0,0 +1,50 @@
+
+#### Arguments
+
+1. `props` (*String* | *String[]*): Defaults to `[all]`. Provides a css property, or a list of css properties that should be transitioned.
```
```suggestion
1. `props` (*String* | *String[]*): Defaults to `['all']`. Provides a css property, or a list of css properties that should be transitioned.
```

oliviertassinari commented on docs/src/pages/customization/transitions/transitions.md:
````diff
@@ -0,0 +1,53 @@
+
+<p class="description">Use the <code>theme.transitions.create()</code> helper to create consistent transitions on the elements of your UI.</p>
+
+```jsx
````
```suggestion
```js
```

oliviertassinari commented on docs/src/pages/customization/transitions/transitions.md:
````diff
@@ -0,0 +1,53 @@
+```js
+const theme = createMuiTheme({
+  transitions: {
+    duration: {
+      standard: 500,
+    },    
````
I think that we can mention all the possible values

```suggestion
    duration: {
      shortest: 150,
      shorter: 200,
      short: 250,
      // most basic recommended timing
      standard: 300,
      // this is to be used in complex animations
      complex: 375,
      // recommended when something is entering screen
      enteringScreen: 225,
      // recommended when something is leaving screen
      leavingScreen: 195,
    },
```

oliviertassinari commented on docs/src/pages/customization/transitions/transitions.md:
```diff
@@ -0,0 +1,53 @@
+const theme = createMuiTheme({
+  transitions: {
+    easing: {
+      sharp: 'cubic-bezier(0.4, 0, 0.8, 1)',
```
I think that we can mention all the easing curves:

```suggestion
      // This is the most common easing curve.
      easeInOut: 'cubic-bezier(0.4, 0, 0.2, 1)',
      // Objects enter the screen at full velocity from off-screen and
      // slowly decelerate to a resting point.
      easeOut: 'cubic-bezier(0.0, 0, 0.2, 1)',
      // Objects leave the screen at full velocity. They do not decelerate when off-screen.
      easeIn: 'cubic-bezier(0.4, 0, 1, 1)',
      // The sharp curve is used by objects that may return to the screen at any time.
      sharp: 'cubic-bezier(0.4, 0, 0.6, 1)',
```

mbrookes commented on docs/src/pages/customization/transitions/transitions.md:
```diff
@@ -0,0 +1,71 @@
+
+## Custom duration
+
+You can change some of the durations by providing a number.
```
```suggestion
You can change some or all of the duration values, or provide your own (for use in the `create()` helper). This example uses the default values which are in milliseconds. You only need to provide the keys you wish to change or add a value for. 
```

(The last sentence might be stating the obvious, but you'd be surprised!)

mbrookes commented on docs/src/pages/customization/transitions/transitions.md:
```diff
@@ -0,0 +1,71 @@
+
+## Custom easing
+
+You can change some of the easing values by providing a custom <code>transition-timing-function</code> value.
```
```suggestion
You can change some or all of the easing values, or provide your own, by providing a custom CSS <code>transition-timing-function</code> value.
```

mbrookes commented:
Should we also cross-reference the Transition components section?

mnajdova commented on docs/src/pages/customization/transitions/transitions.md:
```diff
@@ -0,0 +1,53 @@
+# Transitions
+
+<p class="description">Use the <code>theme.transitions.create()</code> helper to create consistent transitions on the elements of your UI.</p>
```
Ok, let me restructure it a bit, by starting with the customization options for the transitions (and mentioning that they will affect the components that use them). Then I will move to the `duration`, `easing`, `getAutoHeightDuration` and in the end the `create` API.

For the `getAutoHeightDuration` is this description good:
- "Returns the automatic duration based on the provided height"

mnajdova replied:
Address all comments, please take another look

mbrookes commented on docs/src/pages/customization/transitions/transitions.md:
```diff
@@ -10,7 +10,7 @@ theme.transitions.create(['background-color', 'transform']);
 ## Custom duration
 
-You can change some of the durations by providing a number.
+You can change some or all of the duration values, or provide your own (for use in the `create()` helper). This example uses the default values which are in milliseconds. You only need to provide the keys you wish to change or add a value for. 
```
```suggestion
You can change some or all of the duration values, or provide your own (for use in the `create()` helper). This example uses the default values which are in milliseconds. You only need to provide the keys you wish to change or add a value for.
```
Oops, sorry, trailing space.

mbrookes commented on docs/src/pages/customization/transitions/transitions.md:
```diff
@@ -54,6 +48,36 @@ const theme = createMuiTheme({
 
+## `getAutoHeightDuration`
+
+You can customize the default function for calculating the duration based on the height of the elments (useful for collapsing/expanding) components.
```
```suggestion
You can customize the default function for calculating the duration based on the height of the element (useful for collapsing/expanding transitions).
```

mbrookes commented on docs/src/pages/customization/transitions/transitions.md:
```diff
@@ -54,6 +48,36 @@ const theme = createMuiTheme({
+
+## `create`
+
+Use the <code>theme.transitions.create()</code> helper to create consistent transitions on the elements of your UI.</p>
```
```suggestion
Use the <code>theme.transitions.create()</code> helper to create consistent transitions for the elements of your UI.</p>
```

oliviertassinari commented:
> However since the primary focus is customization, I wonder if create() should be documented after duration and easing?

@mbrookes `create` seems to be what most people are interested into: #16928. I have personally never customized the values of the transition but used the transition helper quite often. I would vote for having the `create` helper documented first.

mbrookes commented:
I'd forgotten that we'd long since abandoned putting the page content in any kind of logical order. 😉 Sorry @mnajdova!

Should the examples sit with the respective API sections, as with https://material-ui.com/customization/breakpoints/#api?

How likely is it that someone would want to customise `getAutoHeightDuration()`?

mnajdova commented:
> I'd forgotten that we'd long since abandoned putting the page content in any kind of logical order. 😉 Sorry @mnajdova!
>
> Should the examples sit with the respective API sections, as with https://material-ui.com/customization/breakpoints/#api?

I think this makes sense, I moved the examples together with the API sectoin, and added examples for the create function before the one for `getAutoHeightDuration`.

Now we have this order:
- Durations
- Easings
- API 
  - create
  - getAutoHeightDuration
- References

If we want to have the create function first, we can make the API section first, although it may be a bit inconsistent with the other customization guides...

> How likely is it that someone would want to customise getAutoHeightDuration()?

To be honest I really don't see why someone would, unless they really know what they are doing, or they have very specific timings for the expanding/collapsing animations.. I would never customize it, but I would definitely like to use the create function, so I think that @oliviertassinari is right, it may make most sense if we add that first on the page..

So the alternative sorting proposal is:
- API 
  - create
  - getAutoHeightDuration
- Durations
- Easings
- References

Let me know what you think @mbrookes

mbrookes commented:
> To be honest I really don't see why someone would

Me neither, so I think we can drop that example.

> the alternative sorting proposal is

That's probably the best compromise.

mnajdova commented:
@mbrookes done 👍

mbrookes commented on docs/src/pages/customization/transitions/transitions.md:
```diff
@@ -2,6 +2,40 @@
+
 ## Durations
 
 You can change some or all of the duration values, or provide your own (for use in the `create()` helper). This example uses the default values which are in milliseconds. You only need to provide the keys you wish to change or add a value for.
```
```suggestion
You can change some or all of the duration values, or provide your own for use in the `create()` helper. This example uses the default values, which are in milliseconds. You only need to provide the keys you wish to change or add.
```

mbrookes replied:
Still not totally happy with that, but I think it's a little better than the previous rewording.

mbrookes commented on docs/src/pages/customization/transitions/transitions.md:
````diff
@@ -0,0 +1,86 @@
+```js
+theme.transitions.create(['background-color', 'transform']);
+```
+
````
```suggestion

#### Example

```

mbrookes commented:
🎉

Floriferous commented:
Did this page disappear in the last v4 version? I'm on 4.11.1 and there is no Transitions page under Customization > Theming > XXX

oliviertassinari commented:
@Floriferous https://next.material-ui.com/customization/transitions/

Floriferous commented:
I’m referring to the v4 docs. I don’t know if v5 docs are relevant or not for v4 given all the styling changes. 

> On 27 Apr 2021, at 14:36, Olivier Tassinari ***@***.***> wrote:
> 
> ﻿
> @Floriferous https://next.material-ui.com/customization/transitions/
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

oliviertassinari commented:
@Floriferous v4 has been almost frozen for a year now. We didn't change this API
